client/my-sites/media-library/list.jsx

```diff
@@ -182,61 +182,183 @@ export const MediaLibraryList = React.createClass( {
+				columnWidth={ this._gridSize }
+				height={ height }
+				onSectionRendered={ this.onSectionRendered }
+				ref={ this.setRef }
+				rowCount={ this._rowCount }
+				rowHeight={ this.getRowHeight }
+				scrollTop={ scrollTop }
+				// Trigger update on change.
+				selectedItems={ this.props.mediaLibrarySelectedItems }
+				// Never show a hortizontal scrollbar.
+				style={ { overflowX: 'hidden' } }
+				width={ width } />
+		);
+	},
+
+	renderSizer: function( { onRowsRendered, registerChild } ) {
+		this._onRowsRendered = onRowsRendered;
+		this._registerChild = registerChild;
+
+		if ( ! this.props.disableHeight ) {
 			return (
-				<ListItem
-					key={ 'placeholder-' + i }
-					style={ this.getMediaItemStyle( itemsVisible + i ) }
-					scale={ this.props.mediaScale } />
+				<AutoSizer>
+					{ ( { height, width } ) => this.renderGrid( {
+						registerChild,
+						height,
+						width
+					} ) }
+				</AutoSizer>
 			);
-		}, this );
+		}
+
+		return (
+			<WindowScroller>
+				{ ( { height, scrollTop } ) => (
+					<AutoSizer disableHeight>
+						{ ( { width } ) => this.renderGrid( {
+							registerChild,
+							height,
+							scrollTop,
+							width
+						} ) }
+					</AutoSizer>
+				) }
+			</WindowScroller>
+		);
 	},
 
-	render: function() {
-		var onFetchNextPage;
+	getRowHeight: function( { index } ) {
+		const item = this._gridItems[ index * this._columnCount ];
+
+		if ( item && 'heading' in item ) {
+			return this.props.headingHeight;
+		}
+
+		return this._gridSize;
+	},
+
+	formatDate: function( date ) {
+		const moment = this.props.moment( date );
+		const today = this.props.moment().startOf( 'day' );
+		const yesterday = today.clone().subtract( 1, 'day' );
+		const oneWeekAgo = today.clone().subtract( 7, 'days' );
+		const oneYearAgo = today.clone().subtract( 1, 'year' );
+
+		if ( today.isSame( moment, 'day' ) ) {
+			return this.props.translate( 'Today' );
+		} else if ( yesterday.isSame( moment, 'day' ) ) {
+			return this.props.translate( 'Yesterday' );
+		} else if ( oneWeekAgo.isBefore( moment, 'day' ) ) {
+			return moment.format( 'dddd' );
+		} else if ( oneYearAgo.isBefore( moment, 'day' ) ) {
+			return moment.format( this.props.translate( '%(month)s %(day)s', {
+				comment: 'Date format.',
+				args: { day: 'D', month: 'MMMM' }
+			} ) );
+		}
 
+		return moment.format( 'LL' );
+	},
+
+	getGridItems: function() {
+		const items = [];
+
+		this.props.media.forEach( ( item, index, media ) => {
+			if ( ! index || ! this.props.moment( media[ index - 1 ].date ).isSame( item.date, 'day' ) ) {
+				const trailing = items.length % this._columnCount;
+
+				if ( trailing ) {
+					items.push( ...Array( this._columnCount - trailing ) );
+				}
+
+				items.push( {
+					heading: this.formatDate( item.date ),
+				} );
+
+				items.push( ...Array( this._columnCount - 1 ) );
+			}
+
+			items.push( item );
+		} );
+
+		if ( ! this.props.media.length ) {
+			items.push( {
```

Comment:
To display a placeholder while loading the first batch of media. I'll add a comment.